### PR TITLE
Add optional account overrides to simulateTransaction

### PR DESCRIPTION
**Pull request by Denommus**
#### Problem

At Neon Labs we want to simulate transactions with custom transaction states.

#### Summary of Changes

We added an optional argument to the `simulateTransaction` RPC with `TransactionOverrides` which that simulation will need. Additionally, we added one more parameter to `get_account_overrides_for_simulation` where we can define additional `AccountOverrides` beside the already existing ones so that we can add the `AccountOverrides` passed via `simulateTransaction`

### CODE REVIEW

File: rpc/src/rpc.rs [hunk 1]
```diff
@@ -3698,6 +3700,7 @@ pub mod rpc_full {
             meta: Self::Metadata,
             data: String,
             config: Option<RpcSimulateTransactionConfig>,
+            account_overrides: Option<AccountOverrides>,
```
Comment (steveluscher) [inline rpc/src/rpc.rs:3703]: Instead of creating another parameter, this should go into `RpcSimulateTransactionConfig`.

**Review by steveluscher — COMMENTED** [review]
If this is something we want to go ahead with – and I have no opinion on whether we should or shouldn't – please add updates to `@solana/web3.js` in this PR.

**Comment by Denommus** [comment]
@steveluscher both comments were addressed

**Comment by CriesofCarrots** [comment]
Hmm, overriding existing account state seems fairly antithetical to the point of transaction simulation. Can you say more about the use-case? (Your problem statement doesn't describe a problem.)

Incidentally, we don't want a `solana-runtime` dependency in any client crates, so that change in this PR would have to be reworked.

**Comment by Denommus** [comment]
@CriesofCarrots I moved `account_overrides` to SDK to prevent dependency on the runtime. As for the discussion, I'll talk with my team to give a specific use case to explain what we need.

**Review by steveluscher — CHANGES_REQUESTED** [review]
Marking this as ‘request changes’ pending more information!

**Comment by Denommus** [comment]
In Neon, we need to replay transactions with historical states. In order to do that we use `simulateTransaction` to replay non-Neon transactions. But in order to replay, we need historical account states in that replaying. We store the account states in our internal database, but we need actually to use them. So in order to do that we implemented this `AccountOverrides` in `simulateTransaction`, so that we can replay old transactions with the old account state.

**Comment by Denommus** [comment]
In more detail:

We want to implement the integration of the solidity contract with Solana programs. For example, a developer can create a Solidity contract, and from that contract, they call the Solana program. The Solidity contract changes some data in the Solana state. Then the Solana program is called, but this needs to emulate how it will work on the changed data. After emulation, the Solidity contract is called again, but with the modified data from the Solana program.

In short, we need a tool to emulate Solana programs. The simplest way to emulate that is with `simulateTransaction`. But in order to do that, we need to be able to use predefined account data in the simulated transaction, so that we can use the modified data from the solidity contract.

File: sdk/src/account_overrides.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    where
+        S: serde::Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(self.accounts.len()))?;
+        for (pubkey, account) in self.accounts.iter() {
+            map.serialize_entry(&format!("{}", pubkey), &AccountDef::from(account.clone()))?;
+        }
+        map.end()
+    }
+}
```
Comment (behzadnouri) [inline sdk/src/account_overrides.rs:78]: Why there are manual implementations for `(de)serialize`?

Also fyi, hashmap serializations are not stable:
https://github.com/bincode-org/bincode/issues/230

Comment (Denommus) [reply sdk/src/account_overrides.rs:78]: I'm not implementing Serialize/Deserialize because of bincode, but so that it can be sent as a JSON in the `simulateTransaction` parameter. It's a manual implementation because I want the hashes to be base58-encoded only for this specific struct, and manually implementing that was the easiest approach I came up with.

Comment (behzadnouri) [reply sdk/src/account_overrides.rs:78]: > I'm not implementing Serialize/Deserialize because of bincode

That is not bincode specific. serde does not have stable hashmap serializations;
I don't know if it matters here or not, though.

For (de)serialization implementation, you wouldn't need any of `AccountOverridesVisitor`, `serialize_map`, etc.
You can just convert `AccountOverrides` into or from `HashMap<String, AccountDef>`, and call `(de)serialize` on the `HashMap`.

**Comment by behzadnouri** [comment]
As a note, accounts-override thing was added as a patch to prevent this exploit:
https://twitter.com/TheCryptoBird/status/1501564093613936648?t=mk5N26qt2RHOHR2_zY9oXg&s=19

This change needs to be closely audited to avoid reintroducing same or a similar exploit.
cc @joncinque who worked on the original patch.

**Comment by steveluscher** [comment]
If you decide to continue this work, know that the JavaScript library has moved to https://github.com/solana-labs/solana-web3.js.

File: sdk/src/account_overrides.rs (hunk 2 above)
Comment (Denommus) [reply sdk/src/account_overrides.rs:78]: @behzadnouri I made it that way so that it is deserialized directly from the JSON and we can get more meaningful errors during the deserialization step, though. I do believe this is a better approach than delaying the conversion of String to Pubkey, which would make it a bit more difficult to report the error back to the deserializer.

**Comment by Denommus** [comment]
@steveluscher does that mean I should have a PR on that repo as well?

**Comment by CriesofCarrots** [comment]
This feels risky to me, and I'm not convinced this is the right use of simulateTransaction, since the point of simulation is to use real-cluster account state. I just discovered a similar proposal in the past, and some discussion: https://github.com/solana-labs/solana/pull/19263#issuecomment-902799782
Seems to me like batch simulation would be more appropriate for your use-case as well.
